EKF2: reset position by fusion

haumarco opened the pull request:
<!--

Thank you for your contribution!

Get early feedback through
- Dronecode Discord: https://discord.gg/dronecode
- PX4 Discuss: http://discuss.px4.io/
- opening a draft pr and sharing the link

-->

### Solved Problem
When the external position reset gets triggered we can now fuse the new position as a "normal" measurement. Therefore the position error (estimate-position on map) gets included in the other states and mainly the wind can be better estimated. For large jumps we still reset the position to ensure we dont induce a really strong wind estimate.
Also adjusted the "set global origin" possibility before takeoff during GNSS denied conditions.


### Alternatives
We could also, reset the position and fuse the "measurement" with a high variance. The estimated position would then perfectly align with the sent measurement position but the effects on the other states would not be as harsh for large position jumps (previous estimate / sent measurement).

dagar commented:
Could you do this depending on the context (state of the filter, provided accuracy, etc) and then either use it as a correction if you can (with an appropriate gate), otherwise fallback to a hard position reset?

sfuhrer commented on src/modules/ekf2/EKF/ekf_helper.cpp:
```diff
@@ -107,37 +107,32 @@ bool Ekf::setEkfGlobalOrigin(const double latitude, const double longitude, cons
-			float current_alt = -_state.pos(2) + gps_alt_ref_prev;
-
+			float z_prev = _state.pos(2);
+			float current_alt = -z_prev + gps_alt_ref_prev;
```
```suggestion
			const float current_alt = -z_prev + gps_alt_ref_prev;
```

sfuhrer commented on src/modules/ekf2/EKF/ekf_helper.cpp:
```diff
@@ -107,37 +107,32 @@ bool Ekf::setEkfGlobalOrigin(const double latitude, const double longitude, cons
 			// determine current z
-			float current_alt = -_state.pos(2) + gps_alt_ref_prev;
-
+			float z_prev = _state.pos(2);
```
```suggestion
			const float z_prev = _state.pos(2);
```

sfuhrer commented on src/modules/ekf2/EKF/ekf_helper.cpp:
```diff
@@ -543,6 +538,11 @@ void Ekf::updateHorizontalDeadReckoningstatus()
 	}
 
+#if defined(CONFIG_EKF2_GNSS)
+	if (!_control_status.flags.in_air && !_params.gnss_ctrl && _control_status.flags.fixed_wing) {
```
Add comment here why we need the fixed-wing flag check here.

haumarco commented on src/modules/ekf2/EKF/ekf.cpp:
```diff
@@ -283,23 +284,61 @@ void Ekf::predictState(const imuSample &imu_delayed)
 	Vector2f pos_corrected = _pos_ref.project(lat_deg, lon_deg) + _state.vel.xy() * dt;
 
-	resetHorizontalPositionTo(pos_corrected, sq(math::max(accuracy, 0.01f)));
+	if (!_control_status.flags.in_air || accuracy < 1.f) {
```
This is for a future change that the accuracy can be set by the user or is dependent on the map zoom. But currently this function is only called through the VEHICLE_CMD_EXTERNAL_POSITION_ESTIMATE command and there the accuracy is always 50...

dagar commented on src/modules/ekf2/EKF/ekf.cpp:
```diff
@@ -283,23 +284,64 @@ void Ekf::predictState(const imuSample &imu_delayed)
 	Vector2f pos_corrected = _pos_ref.project(lat_deg, lon_deg) + _state.vel.xy() * dt;
 
-	resetHorizontalPositionTo(pos_corrected, sq(math::max(accuracy, 0.01f)));
+	if (!_control_status.flags.in_air || accuracy < 1.f) {
```
It wouldn't hurt to check the accuracy is > 0 here.

haumarco replied:
if the accuracy is < 0, `math::max(accuracy, FLT_EPSILON)` will handle it, or what do you mean?

dagar replied:
As a way to verify that the good accuracy (accuracy < 1.f) is real vs slipping through uninitialized or zero'd.

dagar commented on src/modules/ekf2/EKF/ekf.h:
```diff
@@ -802,13 +807,20 @@ class Ekf final : public EstimatorInterface
 
 	void resetVerticalPositionTo(float new_vert_pos, float new_vert_pos_var = NAN);
+	template<typename T>
+	void recordResetStateChange(T &status_pos_change, const T &pos_diff, uint8_t &reset_count, uint8_t &reset_count_prev)
```
To be honest I think using this is more confusing than doing it manually, you have to know both current and prev reset counts are tracked and the internal details.

Shorter term how about starting with more specific helpers?

``` C++
    void recordHorizontalPositionReset(const Vector2f &pos_diff);
```

I'm all for making things generic if we can, but the first priority should be easy to use, hard to misuse.

dagar commented on src/modules/ekf2/EKF/position_fusion.cpp:
```diff
@@ -181,3 +164,11 @@ void Ekf::resetHorizontalPositionToLastKnown()
 	resetHorizontalPositionTo(_last_known_pos.xy(), sq(_params.pos_noaid_noise));
 }
+
+void Ekf::resetHorizontalPositionToExternal(const Vector2f &new_horiz_pos, float horiz_accuracy)
```
I would just do this in place so the INFO message can contain more specific context.

Note you already have INFO both here and before the call.
![image](https://github.com/PX4/PX4-Autopilot/assets/84712/b8db396b-4b3a-4fa2-aa8e-570e9bc016ee)

Your first case is fine (""reset position to external observation"), but the second case you could have more context like "external position correction rejected, resetting". It may or may not be useful, but at a minimum it helps trace the execution path.

dagar commented on src/modules/ekf2/EKF/ekf.cpp:
```diff
@@ -281,23 +281,65 @@ void Ekf::predictState(const imuSample &imu_delayed)
 {
+
 	if (!_pos_ref.isInitialized()) {
+		ECL_INFO("Global position reference not initialized, set home position first");
```
```suggestion
		ECL_INFO("Global position reference not initialized, set home position first");
```

At a minimum this should be a warning, and then actually return a status so that the vehicle_command (top level ekf2) can send an acknowledgement back to the user/GCS. Otherwise this is effectively failing silently from a user perspective.

dagar commented on src/modules/ekf2/EKF/ekf.h:
```diff
@@ -281,7 +281,12 @@ class Ekf final : public EstimatorInterface
-		return (!_horizontal_deadreckon_time_exceeded && !_control_status.flags.fake_pos);
+#if defined(CONFIG_EKF2_GNSS)
+		// if there's GPS module but disabled, we want to slack the fake pos check
+		return !_horizontal_deadreckon_time_exceeded && (!_control_status.flags.fake_pos || !_params.gnss_ctrl);
```
Could you expand on the problem you needed to solve here?

If anything could we instead change the deadreckoning conditions contextually instead?

Also keep in mind the gnss_ctrl parameter isn't telling us if we have a module or not.

haumarco replied:
This was something i added with a slightly different approach I tried. But you're correct its not necessary since the deadreckon time never exceeds while having fixed wing and gps disabled and therefore the fake gps does not get triggered.

This has the slight downside that we only have zero velocity / gyro updates but since we dont have any knowledge about local position this should be fine.

dagar commented on src/modules/ekf2/EKF/ekf_helper.cpp:
```diff
@@ -603,6 +597,13 @@ void Ekf::updateHorizontalDeadReckoningstatus()
 
+#if defined(CONFIG_EKF2_GNSS)
+	// for fixed wing, deadreckon time starts after takeoff. Gives enough time for wind deadreckon to be activated
+	if (!_control_status.flags.in_air && !_params.gnss_ctrl && _control_status.flags.fixed_wing) {
```
It seems wrong to rely on `_params.gnss_ctrl` here as well.

Rather than forcing/faking `_time_last_horizontal_aiding` could we instead do it for real preflight via the ZUPT or perhaps fake_pos fusion?

haumarco commented:
I added two things:
1. When inertial dead reckoning is true, there was already no aid data for no_aid_timeout_max. Therefore we need to subtract the timeout from the _time_last_horizontal_aiding.

2. In the beginning, the starting condition was satisfied for the fake-pos fusion becaue horizontal aiding time is exceeded. I added the _horizontal_deadreckon_time_exceeded to the continuing condition (now same as starting condition). This does not change the logic since isHorizontalAidingActive directly disables the fake pos as soon as there are any external measurements.

dagar commented on src/modules/ekf2/EKF/ekf_helper.cpp:
```diff
@@ -643,14 +643,16 @@ void Ekf::updateHorizontalDeadReckoningstatus()
 	} else {
-		_time_last_horizontal_aiding = _time_delayed_us;
+		if (_time_delayed_us > _params.no_aid_timeout_max) {
+			_time_last_horizontal_aiding = _time_delayed_us - _params.no_aid_timeout_max;
```
Did you find this was still necessary? I didn't think it was because isTimedOut() is handling when it's zero and now checking the elapsed time without subtraction.

haumarco replied:
Define necessary ;)
If we want that _valid_timeout_max_ defines the maximum time where we're using inertial dead reckoning then yes. Because after the last measurement, the system is already dead reckoning with IMU data, but we only label it as such after _no_aid_timeout_max_, which is correct.

The question is:
Where does one define the starting point of the inertial dead reckoning? After the last sensor-measurement or after the last sensor timed out.

dagar reviewed (approved):
Needs testing, but otherwise looks good to go from my perspective.
